Fixes #4450 - authorize the tasks

iNecas opened the pull request:
There are two permissions: view_foreman_tasks for viewing and 
edit_foreman_tasks for controlling them.

Two default roles are pre-seeded: Tasks Reader and Tasks Manager. Also the
Anonymous role is added a permission to see user's own tasks, to prevent the
awkward moment when a user triggers a task and can't see the progress or
status.

iNecas commented:
Run the PR test manually (this PR should make also the master passing) http://ci.theforeman.org/job/test_plugin_pull_request/369/

iNecas commented:
There were still some tests failing, here is the green one http://ci.theforeman.org/job/test_plugin_pull_request/374/

iNecas commented:
Current tests http://ci.theforeman.org/job/test_plugin_pull_request/384/

pitr-ch commented on app/controllers/foreman_tasks/tasks_controller.rb:
```diff
@@ -62,8 +62,27 @@ def controller_name
+      else
+        super
+      end
+    end
```
@ares @iNecas mentioned that this may not be needed. How can we get rid of it? (I needed to add it too in foreman_deployments.)

iNecas replied:
http://projects.theforeman.org/issues/8947

pitr-ch commented on app/models/foreman_tasks/concerns/action_triggering.rb:
```diff
@@ -73,6 +73,7 @@ def plan_destroy_action
       # of planning phase is expected to be commited when execution occurs. Also
       # we want to be able to rollback the whole db operation when planning fails.
       def plan_action(action_class, *args)
+        return if ForemanTasks.dynflow.config.disable_active_record_actions
```
nitpick: Isn't `enable_active_record_actions` with default `true` more natural?

iNecas replied:
I've put it this way as it represents better (for me) the default behaviour. It's more straight here, that the active record actions are enabled by default and you can disable it when you want.

Not strong preference about it, so I can change it if you insist :)

pitr-ch replied:
I do not dare to insist on this nit :)

pitr-ch commented on app/models/foreman_tasks/task.rb:
```diff
@@ -83,18 +86,28 @@ def self.search_by_generic_resource(key, operator, value)
+
       key_name = self.connection.quote_column_name(key.sub(/^.*\./,''))
-      joins = <<-JOINS
+      joins = <<-JOIN_OWNER
```
if `<<-SQL` is used instead it will allow certain editors (Rubymine) to highlight the SQL syntax.

iNecas replied:
You have it

pitr-ch replied:
Thaks

pitr-ch replied:
Wow even Github recognized it :)

iNecas replied:
Nice:)

iNecas commented:
Current tests http://ci.theforeman.org/job/test_plugin_pull_request/

iNecas commented:
Current tests http://ci.theforeman.org/job/test_plugin_pull_request/406/

iNecas commented:
@pitr-ch @ares I've refactored the console authorization code a bit and added tests. I'm done with changes on this PR other than addressing PR comments: ready for merge from my perspective

pitr-ch commented on lib/foreman_tasks/dynflow/console_authorizer.rb:
```diff
@@ -0,0 +1,56 @@
+      if @user.filters.joins(:filterings).unlimited.where('filterings.permission_id' => edit_permission).first
+        return true
+      end
+    end
```
nit: this function returns `true` or `nil`

pitr-ch commented:
ACK

ares commented on lib/foreman_tasks/dynflow/console_authorizer.rb:
```diff
@@ -0,0 +1,56 @@
+
+    def allow?
+      if @user && (unlimited_edit? || authorized_for_task?)
+        return true
```
WAT?

ares commented on lib/foreman_tasks/dynflow/console_authorizer.rb:
```diff
@@ -0,0 +1,56 @@
+      if @user && (unlimited_edit? || authorized_for_task?)
+        return true
+      else
+        return false
```
WAT? 2/2

pitr-ch replied:
:D

iNecas replied:
Just to explain how that come to be there (in case you wondered), this is how the method looked before, so I missed the last refactoriing step. Thanks for notifying this. Fixed now

```ruby
def authorized_console?(env)
  rack_request        = Rack::Request.new(env)
  user_id, expires_at = rack_request.session.values_at('user', 'expires_at')

  return false unless Setting[:dynflow_enable_console]

  if Setting[:dynflow_console_require_auth]
    return false if Time.now.to_i > expires_at.to_i
    user = User.find_by_id(user_id)
    return false if user.nil?
    return true if user.admin?

    # users with unlimited edit_foreman_tasks can operate with the
    # console no matter what task it is…
    edit_permission = Permission.where(:name => :edit_foreman_tasks, :resource_type => ForemanTasks::Task.name).first
    if user.filters.joins(:filterings).unlimited.where('filterings.permission_id' => edit_permission).first
      return true
    end

    # limited users can operate only on specific tasks they are authorized to
    dynflow_id = rack_request.path_info[/^\/([\da-f]{8}-[\da-f]{4}-[\da-f]{4}-[\da-f]{4}-[\da-f]{12})/,1]
    unless dynflow_id.empty?
      task = ForemanTasks::Task::DynflowTask.where(:external_id => dynflow_id).first
    end
    if task
      begin
        User.current = user
        return Authorizer.new(user).can?(:edit_foreman_tasks, task)
      ensure
        User.current = nil
      end
    else
      return false
    end
  else
    return true
  end
end
```

pitr-ch replied:
I thought as much, I did not dare to think that you wrote it ;)

iNecas commented:
Current tests http://ci.theforeman.org/job/test_plugin_pull_request/415/
